## Assignment 4
### CS156, Fall 2019
#### Berfin Karaman

In [160]:
#importing the libraries
from sklearn.metrics import accuracy_score
import time
from sklearn import svm
from sklearn.datasets import fetch_openml
import numpy as np
import warnings
warnings.filterwarnings("ignore")
from sklearn.model_selection  import train_test_split as split

In [172]:
#fetching the data from the fetch_openml

mnist = fetch_openml('mnist_784', version=1, cache=True)

x_set = mnist["data"]
y_set = mnist["target"]


In [173]:
print(x_set.shape)
print(y_set.shape)

(70000, 784)
(70000,)


In [174]:
#chosing the digits 7 and 1
digits = np.where((y_set == "1") | (y_set == "7"))
x_set, y_set = x_set[digits], y_set[digits]


In [175]:
#since there is  15170 data points, I will use the half  due to computational limitations (explained in
# evaluation)

x_set = x_set[:7000]
y_set=y_set[:7000]


In [176]:
#splitting the data
x_train, x_test, y_train, y_test = split(x_set, y_set, test_size=0.2)


In [177]:
#Train a support vector classifier with linear kernel
#creating a for loop to try different C values to understand which value gives more accurate results

for i in [0.1, 1, 10]:
    clf = svm.SVC(kernel='linear', C=i) 
    initial = time.time()
    clf.fit(x_train, y_train)
    final = time.time()
    print("Training time for linear kernel when C =", i ,":", '%.2f'%(final - initial), "sec")  

    #error rate
    pred = clf.predict(x_test)
    err_rate = 1 - accuracy_score(y_test, pred)
    print("Error rate for linear kernel when C =", i ,"is", err_rate)

Training time for linear kernel when C = 0.1 : 0.78 sec
Error rate for linear kernel when C = 0.1 is 0.004285714285714337
Training time for linear kernel when C = 1 : 0.79 sec
Error rate for linear kernel when C = 1 is 0.004285714285714337
Training time for linear kernel when C = 10 : 0.80 sec
Error rate for linear kernel when C = 10 is 0.004285714285714337


In [178]:
#Train a support vector classifier with polynomial kernel
#creating a for loop to try different C values and degrees of polynomial
#to understand which value gives more accurate results

for i in [0.1, 1, 10]:
    for j in [1, 2, 3, 4, 5]:
        clf = svm.SVC(kernel='poly', C=i, degree=j) 
        initial = time.time()
        clf.fit(x_train, y_train)
        final = time.time()
        print("Training time for poly kernel when C =", i ,"and degree =",j, ":", '%.2f'%(final - initial), "sec")  

        #error rate
        pred = clf.predict(x_test)
        err_rate = 1 - accuracy_score(y_test, pred)
        print("Error rate for poly kernel when C =", i ,"and degree =",j,"is", err_rate)

Training time for poly kernel when C = 0.1 and degree = 1 : 0.81 sec
Error rate for poly kernel when C = 0.1 and degree = 1 is 0.004285714285714337
Training time for poly kernel when C = 0.1 and degree = 2 : 1.02 sec
Error rate for poly kernel when C = 0.1 and degree = 2 is 0.0035714285714285587
Training time for poly kernel when C = 0.1 and degree = 3 : 1.19 sec
Error rate for poly kernel when C = 0.1 and degree = 3 is 0.004285714285714337
Training time for poly kernel when C = 0.1 and degree = 4 : 1.36 sec
Error rate for poly kernel when C = 0.1 and degree = 4 is 0.007857142857142896
Training time for poly kernel when C = 0.1 and degree = 5 : 1.49 sec
Error rate for poly kernel when C = 0.1 and degree = 5 is 0.011428571428571455
Training time for poly kernel when C = 1 and degree = 1 : 0.79 sec
Error rate for poly kernel when C = 1 and degree = 1 is 0.004285714285714337
Training time for poly kernel when C = 1 and degree = 2 : 0.96 sec
Error rate for poly kernel when C = 1 and degree

In [179]:
#Train a support vector classifier with the RBF kernel
#creating a for loop to try different C values and gamma values
#to understand which value gives more accurate results

for i in [0.1, 1, 10]:
    for j in [0.01, 0.1, 1, 10]:
        clf = svm.SVC(kernel='rbf', C=i, gamma=j) 
        initial = time.time()
        clf.fit(x_train, y_train)
        final = time.time()
        print("Training time for rbf kernel when C =", i ,"and gamma =",j, ":", '%.2f'%(final - initial), "sec")  

        #error rate
        pred = clf.predict(x_test)
        err_rate = 1 - accuracy_score(y_test, pred)
        print("Error rate for rbf kernel when C =", i ,"and gamma =",j,"is", err_rate)

Training time for rbf kernel when C = 0.1 and gamma = 0.01 : 29.09 sec
Error rate for rbf kernel when C = 0.1 and gamma = 0.01 is 0.48571428571428577
Training time for rbf kernel when C = 0.1 and gamma = 0.1 : 31.13 sec
Error rate for rbf kernel when C = 0.1 and gamma = 0.1 is 0.48571428571428577
Training time for rbf kernel when C = 0.1 and gamma = 1 : 29.96 sec
Error rate for rbf kernel when C = 0.1 and gamma = 1 is 0.48571428571428577
Training time for rbf kernel when C = 0.1 and gamma = 10 : 30.17 sec
Error rate for rbf kernel when C = 0.1 and gamma = 10 is 0.48571428571428577
Training time for rbf kernel when C = 1 and gamma = 0.01 : 29.03 sec
Error rate for rbf kernel when C = 1 and gamma = 0.01 is 0.48571428571428577
Training time for rbf kernel when C = 1 and gamma = 0.1 : 29.51 sec
Error rate for rbf kernel when C = 1 and gamma = 0.1 is 0.48571428571428577
Training time for rbf kernel when C = 1 and gamma = 1 : 31.53 sec
Error rate for rbf kernel when C = 1 and gamma = 1 is 0.

#### Evaluation of the results:

Firstly, I wanted to use the whole dataset for the digits 1 and 7, which had 15170 data points. However, when I tried to run the code, it took around 5 minutes to run each iteration of the support vector classifier with the RBF kernel. Thus, I decided to use a subset of the original data to have printed results for each iteration. When I run the whole data set, the polynomial kernel with degree 3 was the best performing one with 0.0029 error rate and 4 seconds training time. However, I did not run all the iterations for the RBF kernel; thus, I would need to rerun the code with the subset data. I will have 7000 data points for the subset data.

Based on the results on the error rate (0.0035 error rate), we can conclude that the support vector classifier with a polynomial kernel and degree 2 performs better than the other models. Additionally, the training time that is required for the polynomial kernel with degree 2 is around 1 second, which is only 0.2 seconds more than the best training time that we received (0.78 with the linear kernel). Since it is advantageous for both error rate and training timing, we can confidently say that in this scenario, the polynomial kernel with degree 2 is more favorable when we use approximately half of the dataset. I observed that the best performing polynomial had a lower higher degree when we used half of the dataset. This makes sense because when we had a bigger data set, we might need a slightly more complex kernel for SVC.   

The SVC with the RBF kernel performed the worst because it has an error rate of around 0.48 and in general, took 30 times more training time than the linear and polynomial kernel. One of the reasons why the RBF kernel performed poorly might be that the RBF kernel is more suitable for non-linearly separable data. Since we had 0.004 error rate with the linear kernel, we can say that the dataset we have is fairly linearly separable.  


